### PR TITLE
Migrate pool to new synchronous API

### LINKED ISSUE #205: Refactor Pool to Reduce Nested Locking
In our current Pool implementation, we rely heavily on nested locking. While this is functionally correct and provides the necessary atomic guarantees, it does not scale well under very high-throughput workloads. In such scenarios, contention increases significantly, along with the risk of lock poisoning. We recently migrated part of the tproxy implementation to use DashMap (a concurrent hash map), and the performance improvement was substantial.

Going forward, we should refactor the Pool to adopt more appropriate concurrent data structures and reduce our dependence on deeply nested locks.

### PR DESCRIPTION
**Pull request by bit-aloo**
closes: #205

### CODE REVIEW

**Comment by bit-aloo** [comment]
Opening this up for review. Let me know what you think of API surface and if anything can be improved or is not clear

File: pool-apps/pool/src/lib/channel_manager/mod.rs [hunk 1]
```diff
@@ -116,7 +90,24 @@ impl ChannelManagerIo {
...
+    downstream_id_factory: Arc<AtomicUsize>,
+    // Mapping of `(downstream_id, channel_id)` -> vardiff controller.
+    // Each entry manages variable difficulty for a specific downstream channel.
+    pub(crate) vardiff: SharedMap<VardiffKey, VardiffState>,
+    // Coinbase outputs.
+    pub(crate) coinbase_outputs: Vec<u8>,
+    // Last new prevhash.
+    pub(crate) last_new_prev_hash: SharedLock<Option<SetNewPrevHash<'static>>>,
+    // Last future template.
+    pub(crate) last_future_template: SharedLock<Option<NewTemplate<'static>>>,
```
Comment (GitGab19) [inline pool-apps/pool/src/lib/channel_manager/mod.rs:105]: Since now these two are independent, we should have a single wrapper behind a single `SharedLock`, containing both of them.

In this way we'll be sure that `last_future_template` and `last_new_prev_hash` are always in sync between each-other.

**Comment by GitGab19** [comment]
My clanker found a real concern about the usage of `self.downstream.get_cloned()` in the channel openings handlers:
```
Possible interleaving:

1. Open-channel handler clones the `Downstream`.
2. The downstream disconnects.
3. `remove_downstream()` removes the downstream and retains away existing `vardiff` entries.
4. The open-channel handler continues with the cloned handle.
5. It inserts a channel into `downstream.standard_channels` / `extended_channels`.
6. It inserts a new `(downstream_id, channel_id)` entry into `self.vardiff`.

At that point, `self.vardiff` can contain an entry for a downstream that no longer exists in `self.downstream`. `run_vardiff()` will then skip it forever because `self.downstream.get_cloned(&downstream_id)` returns `None`.

This is not a memory-safety issue, but it does break the manager invariant that `vardiff` entries correspond to live downstream/channel state. I think the final channel insertion and `vardiff` insertion should happen while holding/rechecking the downstream map entry, or otherwise verify membership at commit time.
```

File: pool-apps/pool/src/lib/channel_manager/mod.rs (hunk 1 above)
Comment (bit-aloo) [reply pool-apps/pool/src/lib/channel_manager/mod.rs:105]: Even if we not keep them inside a single wrapper abstraction, they still would be in sync. Like I can individually update them. I am not understanding, how wrapping them together helps? Can you elaborate more?

Comment (GitGab19) [reply pool-apps/pool/src/lib/channel_manager/mod.rs:105]: The point here is not so much about updating their values in sync, but more about reading their values.

If one of the two values gets updated while something else is reading the other, you could end up in a situation where the `last_future_template` and `last_new_prev_hash` you read are not in sync between each-other.

Is it more clear now?

Comment (bit-aloo) [reply pool-apps/pool/src/lib/channel_manager/mod.rs:105]: Hmm, I do get what you are trying to say. Let’s say the task order is this:

```rust
(Prevhash_old, future_template_old)

1. handle_new_template (Prevhash_old, future_template_new)
   // Currently in an inconsistent state, as the future template does not correspond to the prevhash.

2. downstream connects
   // We initialize it with an inconsistent pair of future template and prevhash.
   // But this is okay, because by the time we send OpenChannelSuccess,
   // it will receive a consistent snapshot of prevhash and future_template.

3. handle_new_prevhash (Prevhash_new, future_template_new)
```

We use a common `select!` block in `ChannelManager`, which processes ChannelManager tasks sequentially.

The only times these values are used are:

1. When a downstream connects (having an inconsistent state here is okay, since it will eventually get corrected).
2. When we run a ChannelManager task, which happens sequentially because of the `select!` block, ensuring we do not read things out of order.

But I do get your point, ideally, we should wait for the pair to become consistent before even performing a read operation.

Currently, and even before this refactor, that was not the case, because both of these fields (`last_new_prevhash` and `last_future_template`) are updated through completely different messages, which naturally introduces a time window where they can be inconsistent.

**Comment by bit-aloo** [comment]
> This is not a memory-safety issue, but it does break the manager invariant that `vardiff` entries correspond to live downstream/channel state. I think the final channel insertion and `vardiff` insertion should happen while holding/rechecking the downstream map entry, or otherwise verify membership at commit time.
> ```

Added here: b3dcf5d, thanks.

File: pool-apps/pool/src/lib/channel_manager/mod.rs (hunk 1 above)
Comment (GitGab19) [reply pool-apps/pool/src/lib/channel_manager/mod.rs:105]: As discussed during yesterday's PR Review Club call, we are gonna improve this later, as tracked by https://github.com/stratum-mining/sv2-apps/issues/551

File: pool-apps/pool/src/lib/channel_manager/mod.rs [hunk 2]
```diff
@@ -116,7 +90,24 @@ impl ChannelManagerIo {
 /// to perform message traversal.
 #[derive(Clone)]
 pub struct ChannelManager {
-    pub(crate) channel_manager_data: Arc<Mutex<ChannelManagerData>>,
+    // Mapping of `downstream_id` -> `Downstream` object,
+    // used by the channel manager to locate and interact with downstream clients.
+    pub(crate) downstream: SharedMap<DownstreamId, Downstream>,
```
Comment (plebhash) [inline pool-apps/pool/src/lib/channel_manager/mod.rs]: on multiple occasions while reviewing this PR I read `self.downstream` and think it's a single object, only to find out later that's actually a collection

```suggestion
    pub(crate) downstreams: SharedMap<DownstreamId, Downstream>,
```

**Comment by plebhash** [comment]
I'm a bit concerned with the usage of `get_cloned` on this PR.

I assume it's being introduced as a strategy to optimize against contention from long-held entry guards in `with` and `with_mut`

more specifically, the point raised by @GitGab19 above has not been fully addressed

there are other potential TOCTOU race windows related to similar usage of `get_cloned` in several places in this PR:
- `handle_open_standard_mining_channel` (as already stated)
- `handle_open_extended_mining_channel` (as already stated)
- `handle_update_channel`
- `handle_set_custom_mining_job`
- `handle_new_template`
- `handle_set_new_prev_hash`
- `run_vardiff`

https://github.com/stratum-mining/sv2-apps/commit/b3dcf5da277ea92f55b995f6c2f5c9676a37ebc0 partially mitigates the issue, but open-channel handlers still allow stale cloned downstream handles to commit channel/vardiff state after disconnect... while I think this could be ok to live with as a deliberate tradeoff for this specific scenario, we're still violating invariants and I also don't think we can trivially generalize this pattern to solve all potential TOCTOU race windows listed above

---

overall, we're already swapping the monolithic/nested-lock model for fine-grained `SharedLock` and `SharedMap`, and we're yet to analyze the performance gains coming from this improvement

let's not forget that the goal of #368 was to introduce safe `DashMap`-based primitives that will help us deprecate the unsafe custom `Mutex`... the fact that now we're bypassing those primitives via `get_cloned` feels like we're moving the goal post in a precipitated way

moreover, #529 still needs to be fixed, so even if de-nesting locks + `DashMap` sharding only brings mild/moderate performance improvements, we still lack empirical proof of broader contention issues

so IMHO we should do things one step at a time:

for now, avoid aggressively using `get_cloned` as an over-optimized strategy to reduce contention on this PR, since that is introducing concurrency side-effects for which mitigation is non-trivial and will likely blow up scope

for now, `get_cloned` usage should be restricted to snapshot-style usage only (and the function should be documented accordingly)

if we later see that contention remains an issue, then we should leverage the foundations established by #368 + this PR to address it with a proper strategy

**Comment by bit-aloo** [comment]
The argument that the above suggestion is only partially valid is not entirely sound.

If you look closely, the concern raised above is that we are checking liveness only at the end rather than at every commit point. This is intentional, as I wanted to keep the logic simple and easier to reason about. These intermediate commit points do not materially affect the outcome. If a downstream disconnect occurs during channel opening, performing the cleanup at the end is sufficient. Otherwise, we would need to introduce cleanup logic after every commit step, which would significantly increase complexity and reduce maintainability.

Regarding the TOCTOU concern, where it is mentioned that a race could occur in-between the channel manager methods and the vardiff task, I agree that the concern is partially valid(only for vardiff and select task). However, all channel manager methods execute sequentially, so a race condition cannot occur between those methods themselves. The only concurrency involved is with the vardiff task. A side note the partial commits are ok if the object eventually gonna be dropped

It will be useful, if you can mention the TOCTOU scenarios?

Thanks for review though, would be awesome discuss more on this.

**Comment by plebhash** [comment]
> so a race condition cannot occur between those methods themselves. 

that is not the claim here, it's about `handle_*` execution and `remove_downstream`

> The argument that the above suggestion is only partially valid is not entirely sound.

`is_alive` introduced via b3dcf5d is only partial mitigation because it could return `true` and allow execution to continue, only for `remove_downstream` to be executed immediately afterwards and break invariants again

end of the day, the truth is that with this extensive usage of `get_cloned`, the removal of a `Downstream` object is not atomic, and `is_alive` only gives us loose guarantees but no sound atomicity

if we decided to live with these loose guarantees, we should consciously accept the following tradeoffs:
- potential wasted CPU cycles
- potential misleading logs with concurrent contradictory signals about the application state
- temporary ghost vardiff entries from TOCTOU races on `handle_open_*`

(maybe more, these are the ones I can identify right now)

but the question is: do we want/need to make that decision now?

---

overall I'm not going to die on this hill, and I don't want to spend more time debating all the details of this rabbit hole

I'd rather focus on my broader point, which I'm happy to clarify in case it's not already sufficiently clear:

the extensive usage of `get_cloned` as a strategy for contention optimization feels precipitated (given scopes of #368 + #529), and it seems we don't have a good grasp of the second-order consequences of it, so I don't see the point in rushing things at the expense of a new category of unknown problems to be solved in the future

IMHO we should stick to the originally planned scope and measure the performance gains before we jump into new strategies for contention optimization... if we later find that we're still facing performance issues, we can evaluate `get_cloned` as a potential strategy and its inherent tradeoffs more carefully

**Comment by bit-aloo** [comment]
Hmm, the idea behind introducing the clone was not because of any technical limitation or API issue . It was mainly about making the code easier to write and follow. If you look at the code now, it is just a simple flow: get the object, perform the operation, and move on, let the system follow eventual consistency. 

Even if the object becomes invalid, the operation can still happen, and the system will eventually recover and become consistent again. That was the idea behind this. The introduction of `is_alive` was also along the same lines: making the system faster by relying on eventual consistency instead of always enforcing strong consistency. The side effects of operating on an invalid object were not expected to be serious. At least, that was my thinking.

That said, nothing is stopping us from providing atomic access to objects and ensuring they cannot become invalid while they are being used. That would give us stronger consistency guarantees (like before).

I agree that we should get some numbers before deciding. I will add a commit that prevents operations on invalid objects. Cool.

**Comment by bit-aloo** [comment]
Removed get_cloned for downstream, here: [ce528a0](https://github.com/stratum-mining/sv2-apps/pull/531/commits/ce528a07eeef60b9039d408812321ca263c708f1).


I will squash commit later, would make it easy to review.

**Comment by plebhash** [comment]
> Even if the object becomes invalid, the operation can still happen, and the system will eventually recover and become consistent again. That was the idea behind this. The introduction of is_alive was also along the same lines: making the system faster by relying on eventual consistency instead of always enforcing strong consistency. The side effects of operating on an invalid object were not expected to be serious. At least, that was my thinking.

I think this is a reasonable rationale, and I wouldn't completely rule it out of the design space yet.

but tbh conflicting logs and ghost vardiff entries that could last up to 60s in memory worry me about the UX implications

now that `sv2-ui` is a first-class citizen on the SRI stack, we must always put ourselves in the shoes of the end user before we make key architectural/design decisions

> Hmm, the idea behind introducing the clone was not because of any technical limitation or API issue . It was mainly about making the code easier to write and follow. If you look at the code now, it is just a simple flow: get the object, perform the operation, and move on, let the system follow eventual consistency.

I empathize with the urge to make devX better. It's been painful to live with closure complexity of the soon-to-be-deprecated custom `Mutex`, and I also wish it went away.

But we made it this far, and migrating away from it is a delicate operation. So I feel it's wiser to move slow, evaluate consequences abroad the entire SRI stack, and optimize things gradually.

We laid some solid foundation on #368. Let's take some time and explore it. How will granularly locked sharded maps perform in face of curtailed loads (with granular instrumentation of the source code)? I'm genuinely curious to look at those numbers!

> Removed get_cloned for downstream, here: [ce528a0](https://github.com/stratum-mining/sv2-apps/pull/531/commits/ce528a07eeef60b9039d408812321ca263c708f1).
> 
> I will squash commit later, would make it easy to review.

awesome, thanks!

**Comment by plebhash** [comment]
can we clean up commit history?

e0fe2cb7eade507026278366f1014efe4c95458e and ce528a07eeef60b9039d408812321ca263c708f1 are a bit redundant now (add `get_cloned` then remove `get_cloned`)

**Comment by bit-aloo** [comment]
> I will squash commit later, would make it easy to review.

 I assume you had looked into the final commit earlier.

> can we clean up commit history?
> 
> [e0fe2cb](https://github.com/stratum-mining/sv2-apps/commit/e0fe2cb7eade507026278366f1014efe4c95458e) and [ce528a0](https://github.com/stratum-mining/sv2-apps/commit/ce528a07eeef60b9039d408812321ca263c708f1) are a bit redundant now (add `get_cloned` then remove `get_cloned`)

Squashed. Awesome, thanks.

**Review by plebhash — APPROVED** [review]

File: pool-apps/pool/src/lib/monitoring.rs [hunk 3]
```diff
@@ -109,12 +107,8 @@ impl Sv2ClientsMonitoring for ChannelManager {
...
-        self.channel_manager_data
-            .safe_lock(|d| {
-                d.downstream
-                    .get(&client_id)
-                    .and_then(downstream_to_sv2_client_info)
-            })
-            .unwrap_or(None)
+        self.downstreams
+            .get_cloned(&client_id)
+            .and_then(|downstream| downstream_to_sv2_client_info(&downstream))
```
Comment (GitGab19) [inline pool-apps/pool/src/lib/monitoring.rs]: Why are you still using `get_cloned()` here?

According to [your comment](https://github.com/stratum-mining/sv2-apps/pull/531#issuecomment-4691121816), you should have removed it?

Comment (bit-aloo) [reply pool-apps/pool/src/lib/monitoring.rs]: Slipped, thanks updated.

File: stratum-apps/src/sync.rs [hunk 4]
```diff
@@ -122,6 +123,16 @@ impl<K: Eq + Hash + Clone, V> SharedMap<K, V> {
         SharedMap(Arc::new(DashMap::new()))
     }
 
+    /// Get an owned clone of a value.
+    ///
+    /// Prefer [`Self::with`] when only part of a large value is needed.
```
Comment (plebhash) [inline stratum-apps/src/sync.rs]: let's make docs explicitly clear about the appropriate vs inappropriate usage of this method

File: stratum-apps/src/sync.rs
Comment (bit-aloo) [inline stratum-apps/src/sync.rs:237]: I have a question about the `for_each` APIs. In cases like this, if an error occurs, we simply return. Since we're iterating over a collection, multiple items may fail, and we generally don't want a single error to cause an early exit and prevent the remaining items from being processed.

At the same time, I would prefer not to make the API surface significantly more complex. However, I think we do need some form of batching or error aggregation here to ensure that iteration isn't aborted prematurely. 

All ears for new design.

Comment (GitGab19) [reply stratum-apps/src/sync.rs:237]: My clanker gave me this example from `tokio`, which could be taken into consideration as a possible approach: https://docs.rs/tokio/latest/tokio/task/struct.JoinSet.html#method.join_all

In the `join_all` docs, they explicitly say: "If any tasks on the JoinSet fail with an [JoinError](https://docs.rs/tokio/latest/tokio/task/struct.JoinError.html), then this call to join_all will panic and all remaining tasks on the JoinSet are cancelled. To handle errors in any other way, manually call [join_next](https://docs.rs/tokio/latest/tokio/task/struct.JoinSet.html#method.join_next) in a loop.`

The `join_next` docs can be found here: https://docs.rs/tokio/latest/tokio/task/struct.JoinSet.html#method.join_next

---

In our case here, the method corresponding to the tokio's `join_next()` is the `for_each()` that we already have, right?

So I guess we can use that when we know there's a fallible operation, and collect the eventual errors at the caller site?

Maybe I'm missing something?

**Comment by jokeez** [comment]
Hi — reviewed this in context of #551/#552 (TDP cache/snapshot ordering).

I can contribute a small state-transition test matrix for:
- NewTemplate -> SetNewPrevHash ordering
- missing/late SetNewPrevHash after NewTemplate
- stale snapshot reuse during channel init

No commercial pitch — just reproducible edge cases + expected invariants.
If useful, I’ll post it here as a markdown table first, then optional PR with tests.

File: stratum-apps/src/sync.rs
Comment (plebhash) [reply stratum-apps/src/sync.rs:237]: @Shourya742 is there some part of this PR that you feel is blocked by this?

I'm asking because we already have #182 (batch error flow) and #257 (batched disconnect)

I'm not against brainstorming here, this is a good opportunity because in a way we're still trimming the rough edges of #368

but since we already have trackers for this specific issue, it might make sense to tackle that in a separate PR?

unless there's a concrete blocker for this PR ofc

---

btw, from a relatively superficial analysis I think `next` seems to make sense... instead of looping inside the function body, do it on the call site and handle errors individually (e.g.: disconnect) before moving on to the next iteration

maybe we can take note of `next` pattern as a potential solution on #182 and #257 and proceed with this PR without addressing this specific issue?

Comment (bit-aloo) [reply stratum-apps/src/sync.rs:237]: This sounds good as well, I will update this PR, and we can proceed with the merge.

File: stratum-apps/src/sync.rs (hunk 4 above)
Comment (GitGab19) [reply stratum-apps/src/sync.rs]: Maybe we can add something specifically about the risks of using it here @Shourya742 ?

Comment (bit-aloo) [reply stratum-apps/src/sync.rs]: I did add a note here: [24afb06](https://github.com/stratum-mining/sv2-apps/pull/531/commits/24afb06fae3044904628de2ea629e0dd5441e533)

Comment (GitGab19) [reply stratum-apps/src/sync.rs]: This is what my clanker would suggest as docs:
```
/// Get an owned clone of a value.
///
/// This releases the map entry guard immediately, so the entry may be removed
/// or replaced while the caller is still using the clone. Use this only when
/// stale clones are acceptable. Prefer [`Self::with`] or [`Self::with_mut`]
/// when later work depends on the entry still being present.
```

Comment (bit-aloo) [reply stratum-apps/src/sync.rs]: done

File: pool-apps/pool/src/lib/channel_manager/mod.rs [hunk 5]
```diff
@@ -699,6 +705,32 @@ impl ChannelManager {
 
         Ok(())
     }
+
+    /// Runs `f` for a downstream that must still be connected.
+    ///
+    /// Returns `DownstreamNotFound` if the downstream disappeared before the lookup completed.
```
Comment (GitGab19) [inline pool-apps/pool/src/lib/channel_manager/mod.rs]:
```suggestion
    /// Runs `f` while holding the downstream map entry guard.
    ///
    /// Use this when mutations must only happen if the downstream is still
    /// registered in the ChannelManager. Do not await, send messages, or re-enter
    /// `self.downstreams` inside `f`.
    ///
    /// Returns the closure result if the downstream is registered. Returns
    /// `DownstreamNotFound` with a disconnect action if the downstream is no
    /// longer registered.
```

File: pool-apps/pool/src/lib/channel_manager/mod.rs [hunk 6]
```diff
@@ -699,6 +705,32 @@ impl ChannelManager {
...
+    #[allow(clippy::result_large_err)]
+    fn with_live_downstream<R, F>(
```
Comment (GitGab19) [inline pool-apps/pool/src/lib/channel_manager/mod.rs]:
```suggestion
    fn with_registered_downstream<R, F>(
```

File: pool-apps/pool/src/lib/channel_manager/mod.rs (hunk 5 above)
Comment (bit-aloo) [reply pool-apps/pool/src/lib/channel_manager/mod.rs]: compiler wont allow using .awaits inside non async closure. I think we are kinda ok even if we not mention the await point

Comment (GitGab19) [reply pool-apps/pool/src/lib/channel_manager/mod.rs]: Alternative without the `async` word :laughing: :
```
/// Runs `f` while holding the downstream map entry guard.
///
/// Use this when mutations must only happen if the downstream is still
/// registered in the ChannelManager. Keep `f` short: do not perform blocking
/// work, send/forward messages, or re-enter `self.downstreams` inside it.
///
/// Returns the closure result if the downstream is registered. Returns
/// `DownstreamNotFound` with a disconnect action if the downstream is no
/// longer registered.
```

File: pool-apps/pool/src/lib/channel_manager/mining_message_handler.rs [hunk 7]
```diff
@@ -78,27 +73,21 @@ impl HandleMiningMessagesFromClientAsync for ChannelManager {
...
-            })
+        self.with_live_downstream(downstream_id, |downstream| {
+            downstream
+                .group_channel
+                .with(|group_channel| {
+                    if group_channel.has_channel_id(msg.channel_id) {
+                        group_channel.remove_channel_id(msg.channel_id);
+                    }
+                })
+                .map_err(PoolError::shutdown)?;
```
Comment (GitGab19) [inline pool-apps/pool/src/lib/channel_manager/mining_message_handler.rs:84]: Why did you introduce this part about group channel?

Comment (bit-aloo) [reply pool-apps/pool/src/lib/channel_manager/mining_message_handler.rs:84]: because we were not cleaning it before when a close channel message is received from dowstream. I had a separate commit for this before to be explicit about this change but got squashed during rebasing.

File: pool-apps/pool/src/lib/channel_manager/mining_message_handler.rs [hunk 8]
```diff
@@ -579,143 +639,209 @@ impl HandleMiningMessagesFromClientAsync for ChannelManager {
...
-                            info!(
-                                "SubmitSharesStandard: valid share | downstream_id: {}, channel_id: {}, sequence_number: {}, share_hash: {}, share_work: {} ✅",
-                                downstream_id, channel_id, msg.sequence_number, share_hash, share_work
-                            );
+                                Ok(messages)
+                            });
+                    match validation {
+                        Some(validation) => {
+                            self.vardiff.with_mut(&vardiff_key, |vardiff| {
+                                vardiff.increment_shares_since_last_update()
```
Comment (GitGab19) [inline pool-apps/pool/src/lib/channel_manager/mining_message_handler.rs:820]: Why did you move the `increment_shares_since_last_update()` inside the `Some()`?

Comment (bit-aloo) [reply pool-apps/pool/src/lib/channel_manager/mining_message_handler.rs:820]: because validation would not happen only in case a channel doesn't exist, so we only increment the vardiff if the channel is actually present. I would say this is mostly because of the datastructure API that we have to have such matches, otherwise this would always be some considering we are already checking for channel existence before.

Comment (GitGab19) [reply pool-apps/pool/src/lib/channel_manager/mining_message_handler.rs:820]: Yeah, I just checked out the PR branch locally, and I verified that everything is still the same.

Sorry but looking at the diffs over github is complicated in this part of the codebase.

File: pool-apps/pool/src/lib/monitoring.rs [hunk 9]
```diff
@@ -12,95 +12,93 @@ use crate::{channel_manager::ChannelManager, downstream::Downstream};
 /// Helper to convert a Downstream to Sv2ClientInfo.
 /// Returns None if the lock cannot be acquired (graceful degradation for monitoring).
```
Comment (GitGab19) [inline pool-apps/pool/src/lib/monitoring.rs]:
```suggestion
```

**Review by GitGab19 — APPROVED** [review]

**Comment by plebhash** [comment]
I feel bad for pointing this out over and over, but I have to ask:

why do we have 0f2dad92b52622ba85a2d7b7e990f835ef4aa570 and 262b54552df86818a5902262553bf251ab3f2d0c as standalone commits?
